src/AWS.Deploy.ServerMode.Client/ServerModeSession.cs

```diff
@@ -0,0 +1,277 @@
+        /// <param name="cancellationToken">A cancellation token that can be used by other objects or threads to receive notice of cancellation.</param>
+        /// <exception cref="InternalServerModeException">Throws when deployment tool server failed to start for un unknown reason.</exception>
+        /// <exception cref="PortUnavailableException">Throws when deployment tool server failed to start due to unavailability of free ports.</exception>
+        Task Start(CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Builds <see cref="IRestAPIClient"/> client using the cached base URL.
+        /// If succeeded, <param name="restApiClient"></param> is initialized with current session client.
+        /// </summary>
+        /// <param name="credentialsGenerator">Func to that provides AWS credentials</param>
+        /// <param name="restApiClient"><see cref="IRestAPIClient"/> client to initialize.</param>
+        /// <returns>True, if <param name="restApiClient"/> is initialized successfully.</returns>
+        bool TryGetRestAPIClient(Func<Task<AWSCredentials>> credentialsGenerator, out IRestAPIClient? restApiClient);
+
+        /// <summary>
+        /// Builds <see cref="IDeploymentCommunicationClient"/> client using the cached base URL.
+        /// If succeeded, <param name="deploymentCommunicationClient"></param> is initialized with current session client.
+        /// </summary>
+        /// <param name="deploymentCommunicationClient"><see cref="DeploymentCommunicationClient"/> client to initialize.</param>
+        /// <returns>True, if <param name="deploymentCommunicationClient"/> is initialized successfully.</returns>
+        bool TryGetDeploymentCommunicationClient(out IDeploymentCommunicationClient? deploymentCommunicationClient);
+
+        /// <summary>
+        /// Returns the status of the deployment server by checking /api/v1/health API.
+        /// Returns true, if the deployment server returns a success HTTP code.
+        /// </summary>
+        /// <param name="cancellationToken">A cancellation token that can be used by other objects or threads to receive notice of cancellation.</param>
+        /// <returns>Returns true, if the deployment server returns a success HTTP code.</returns>
+        ValueTask<bool> IsAlive(CancellationToken cancellationToken);
+    }
+
+    public class ServerModeSession : IServerModeSession, IDisposable
+    {
+        private const int TCP_PORT_ERROR = -100;
+
+        private readonly int _startPort;
+        private readonly int _endPort;
+        private readonly bool _diagnosticLoggingEnabled;
+
+        private string? _baseUrl;
+        private Aes? _aes;
+
+        private string HealthUrl
+        {
+            get
+            {
+                if (_baseUrl == null)
+                {
+                    throw new InvalidOperationException($"{nameof(_baseUrl)} must not be null.");
+                }
+
+                return $"{_baseUrl}/api/v1/health";
```

Comment:
What prevents this url from getting out of sync with what is defined in the rest api?

Reply:
Rest api is generated code and we don't have any way of sharing the code there AFIK. We will need to keep it up to date as we bump the versions of our APIs.